### DISCUSSION TITLE
Microsoft Graph API pagination

### DISCUSSION THREAD

**Discussion by Usama420-git** [grafana/grafana-infinity-datasource#488] [Specific APIs] [answered]
I am using Microsoft Graph API with your plugin and wants to thank you also to giving us this beautiful plugin 

I want to ask I am fetching direct routing calls with https://graph.microsoft.com/v1.0/communications/callRecords/getDirectRoutingCalls(fromDateTime=2019-11-01,toDateTime=2019-12-01) api

 when the record is above 1000 it is give an 
`"@odata.nextLink": "https://graph.microsoft.com/v1.0/communications/callRecords/getDirectRoutingCalls(fromDateTime=2022-11-17,toDateTime=2022-12-14)?$top=10&$skip=1000",` 
to get the record of after 1000 records so is there any thing I can do in the plugin to automatically call nextLink when data is above 1000 rows

![postman](https://user-images.githubusercontent.com/51694624/207560392-020ee6b1-e5f9-49be-abc1-cfa8605b715b.PNG)

**Comment by yesoreyeram** [comment]
Nope.. pagination is not supported by the plugin yet.

**Reply by Usama420-git** [reply]
thank you so much for your answer I hope in the future we'll get pagination in the plugin but for now do you suggest any plugin to me which I can use?

**Reply by yesoreyeram** [reply]
I don't see any plugins for ms graph yet. Curious to know about your usecases for using MS graph inside grafana. That will help to plan this further

**Reply by Usama420-git** [reply]
I am using because we want to see the users presence and direct routing calls like how many calls by date so we can see the graph but the issue is record we are getting is limited to 1000 only

**Comment by ebachman24** [comment, marked as answer]
While pagination isn't necessarily supported by the plugin, you can manually deal with it if you roughly know how many results you're going to produce.  You can create several identical queries calling out different pages, then use a 'Merge' Transformation, which will essentially union it all together.  
i.e.
![image](https://user-images.githubusercontent.com/104096952/208670471-d8b19abc-64c8-49b1-9bfa-561442a8998e.png)

In your case, you're going to need to know roughly how many users you're trying to query, if 2,000 - 3,000, then do 3 queries (1,000 max results each)

Because you are manually defining pages, this method will not scale with your data, which is the biggest downfall.

**Reply by yesoreyeram** [reply]
https://github.com/yesoreyeram/grafana-infinity-datasource/discussions/601

**Comment by ronald-heijn-buys** [comment]
@Usama420-git, I am currently configuring Grafana with this plugin to use with the Microsoft Graph. I followed this https://sriramajeyam.com/grafana-infinity-datasource/wiki/azure-authentication/ documentation and got it to work with Azure Resources. However, can you help me out what settings are required to get it to work with Microsoft Graph?

CC: @yesoreyeram
